### PR TITLE
Query param doc

### PR DESCRIPTION
**Pull request by aeng**
https://zanata.atlassian.net/browse/ZNTA-1730

This would losen the restriction on doc name validation by non-file-endpoint and allow offlinepo to work with File project type.

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/412)

### CODE REVIEW

**Comment by seanf** [comment]
Review status: 0 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-Ko669bFr3CWj12q8dMM:-Ko669bFr3CWj12q8dMN:bb3qpuo) (no related file):*
Just so we don't forget: the client needs to be able to fall back on the old end-point in case the new end-point doesn't exist, ie when a new client connects to an old server. Preferably without hitting the old end-point for every different document in the project (ie it should remember to use the old end-point within a single run).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by huangp** [comment]
Reviewed 3 of 17 files at r1, 19 of 19 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java, line 186 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooBykIHhqpOC5gUCSB:-KooBykIHhqpOC5gUCSC:b-xi59ci) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java#L186)):*
> ```Java
>             Resource resource, @QueryParam("ext") Set<String> extensions,
>             @QueryParam("id") @DefaultValue("") String id,
>             @QueryParam("copyTrans") @DefaultValue("true") boolean copytrans);
> ```

copyTrans is set to false in client (during push). Might take this opportunity to change default value to false here. Or even remove this parameter from the method. People should be using CopyTransResource instead. 

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java, line 188 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooCdUm715GajXA00YT:-KooCdUnwfpZ2Nwu6JYt:b-ldbne3) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java#L188)):*
> ```Java
>     @TypeHint(Resource.class)
>     @Path(RESOURCE_PATH)
>     // /r/resource?
> ```

This seems odd. We have ```project/p/some-project/iteration/i/master``` and then in the end ```/r/resource```. The order is not consistent. We could change the class level PATH to use something without the r and then change all methods to either add ```/r``` (old methods) or use ```/resource``` (new methods)

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java, line 272 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooDUG1ZqYSM4kjKd1L:-KooDUG1ZqYSM4kjKd1M:b-ya81ng) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java#L272)):*
> ```Java
>     @PUT
>     @Path(RESOURCE_PATH)
>     // /r/resource?
> ```

Same as above

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java, line 330 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooDf8Fk0-HfDpzXkuO:-KooDf8Fk0-HfDpzXkuP:b-2guby3) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java#L330)):*
> ```Java
>     @DELETE
>     @Path(RESOURCE_PATH)
>     // /r/resource?
> ```

same here

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java, line 387 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooDkuW1HDRrwWqPzgk:-KooDkuW1HDRrwWqPzgl:b-2guby3) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java#L387)):*
> ```Java
>     @GET
>     @Path(RESOURCE_PATH + "/meta")
>     // /r/resource/meta
> ```

same here

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java, line 451 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooDpd0NbjqLv7_y_aI:-KooDpd0NbjqLv7_y_aJ:b-2guby3) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java#L451)):*
> ```Java
>     @PUT
>     @Path(RESOURCE_PATH + "/meta")
>     // /r/resource/meta
> ```

same here

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/TranslatedDocResource.java, line 147 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooEIQVffzBIWbp_X9U:-KooEIQVffzBIWbp_X9V:b-amfrvo) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/TranslatedDocResource.java#L147)):*
> ```Java
>     @GET
>     @Path(RESOURCE_PATH + "/translations/{locale}")
>     // /r/resource/translations/{locale}
> ```

Might consider changing the path to be consistent

---

*[client/zanata-rest-client/src/main/java/org/zanata/rest/client/AsyncProcessClient.java, line 101 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooGKX4lvx4cPvpQoBT:-KooGKX4lvx4cPvpQoBU:b-juxn4s) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/client/zanata-rest-client/src/main/java/org/zanata/rest/client/AsyncProcessClient.java#L101)):*
> ```Java
>                 .request(MediaType.APPLICATION_XML_TYPE)
>                 .put(Entity.xml(resource));
>         if (RestUtil.isNotFound(response)) {
> ```

I suspect this may cause a "connection not released error" (just checking the status code without consuming the entity/body will not release the connection from memory). Make sure we test this scenario.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java, line 191 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooL_mbrJIE-FPVGjhK:-KooL_mbrJIE-FPVGjhL:bs893h4) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java#L191)):*
> ```Java
>         if (StringUtils.isBlank(id)) {
>             return Response.status(Response.Status.BAD_REQUEST)
>                     .entity("missing id").build();
> ```

For this resource we produce xml and json. Although our client always request xml, if someone else request json, this will generate an marshalling error (a string by itself is not valid json). We might want to create a generic DTO to hold error message and use that instead of plain string everywhere.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/TranslatedDocResourceService.java, line 142 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooM_ayuq9fUfbW7Zm3:-KooM_ayuq9fUfbW7Zm4:b-84v2k1) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/TranslatedDocResourceService.java#L142)):*
> ```Java
>         if (StringUtils.isBlank(id)) {
>             return Response.status(Response.Status.BAD_REQUEST)
>                     .entity("missing id").build();
> ```

Same here. Not json friendly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java, line 186 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooBykIHhqpOC5gUCSB:-KooXiDujp4ZplZJZK8P:b-pg24eb) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java#L186)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

copyTrans is set to false in client (during push). Might take this opportunity to change default value to false here. Or even remove this parameter from the method. People should be using CopyTransResource instead. 
</blockquote></details>

IIRC, that depends on the client version. Changing the default will change the behaviour for old clients which predate this parameter, although it seems unlikely anyone will greatly miss copytrans.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java, line 191 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooL_mbrJIE-FPVGjhK:-KooYQ6B6vTOXsz2Odfl:b-qedn28) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java#L191)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

For this resource we produce xml and json. Although our client always request xml, if someone else request json, this will generate an marshalling error (a string by itself is not valid json). We might want to create a generic DTO to hold error message and use that instead of plain string everywhere.
</blockquote></details>

Unfortunately, all of our entities for failure response are completely unstandardised *strings*. Whether the request Accepts XML or JSON has nothing to do with it.

Basically right now the entity is text for humans; the status code is the only thing machines can rely on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by seanf** [comment]
Reviewed 1 of 19 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java, line 191 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooL_mbrJIE-FPVGjhK:-KooxknWGjVFvQaqbdI7:b-y0io1l) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java#L191)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Unfortunately, all of our entities for failure response are completely unstandardised *strings*. Whether the request Accepts XML or JSON has nothing to do with it.

Basically right now the entity is text for humans; the status code is the only thing machines can rely on.
</blockquote></details>

I agree that we should return an entity, even for failures, which matches the requested type (XML or JSON). Given that the error strings weren't really machine readable before, I don't think backwards compatibility will be an issue.

This looks like a good pattern to follow, in terms of Enunciate annotations for error responses :-)

https://github.com/zanata/zanata-mt/blob/f8845de0fcd3ee05378ffe1d74dc31f2c247402c/server/src/main/java/org/zanata/mt/api/service/BackendResource.java#L37

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by seanf** [comment]
Reviewed 3 of 17 files at r1, 18 of 19 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java, line 175 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-Kop046IG0-f45rm1xFE:-Kop046IG0-f45rm1xFF:bli4rrn) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java#L175)):*
> ```Java
>      *         indicating what happened.<br>
>      *         INTERNAL SERVER ERROR(500) - If there is an unexpected error in
>      *         the server while performing this operation.
> ```

These will probably have to change to `@StatusCodes` and `@ResponseCode` at some point, as used in https://github.com/zanata/zanata-platform/pull/415.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java, line 185 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KoozH1fm3wx1J5oUC8U:-KoozH1fm3wx1J5oUC8V:b7w9fv0) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java#L185)):*
> ```Java
>             @PathParam("iterationSlug") String iterationSlug,
>             Resource resource, @QueryParam("ext") Set<String> extensions,
>             @QueryParam("id") @DefaultValue("") String id,
> ```

Given that we're creating new end points, perhaps we shouldn't call it `id` this time. `docId` is slightly ambiguous for us (given that HDocument has both `id` and `docId`), but I think it's better than `id`. `docId` is also what we use for the similar query param in FileResource.

This applies to all the other new end points.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java, line 186 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooBykIHhqpOC5gUCSB:-Kop-S3NWNy5HkMcdx_Y:b76ufu5) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java#L186)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

IIRC, that depends on the client version. Changing the default will change the behaviour for old clients which predate this parameter, although it seems unlikely anyone will greatly miss copytrans.
</blockquote></details>

Sorry, I forgot this was a new end-point when I replied to Patrick. Yes, the DefaultValue should be false for this one.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java, line 188 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooCdUm715GajXA00YT:-Kop-rUDG9aIqmsQlGDS:bb3svvo) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/SourceDocResource.java#L188)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

This seems odd. We have ```project/p/some-project/iteration/i/master``` and then in the end ```/r/resource```. The order is not consistent. We could change the class level PATH to use something without the r and then change all methods to either add ```/r``` (old methods) or use ```/resource``` (new methods)
</blockquote></details>

Weren't we trying out a new URL scheme for new end-points which was less convoluted and repetitive?

---

*[client/zanata-rest-client/src/main/java/org/zanata/rest/client/AsyncProcessClient.java, line 58 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-Kop0u-r3vuqp3QUZcMn:-Kop0u-r3vuqp3QUZcMo:b-xr9k67) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/client/zanata-rest-client/src/main/java/org/zanata/rest/client/AsyncProcessClient.java#L58)):*
> ```Java
>     @Override
>     // TODO: remove this test when parent method is removed
>     public ProcessStatus startSourceDocCreation(String docId,
> ```

Was the name idNoSlash wrong? Why change it?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java, line 132 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-Kop1u8S7bZ7aZTyFjAl:-Kop1u8S7bZ7aZTyFjAm:b-uou4jh) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/AsynchronousProcessResourceService.java#L132)):*
> ```Java
>             final Resource resource, final Set<String> extensions,
>             final boolean copytrans) {
>         return startSourceDocCreationOrUpdateWithDocId(projectSlug,
> ```

Shouldn't we convert the idNoSlash to a docId before calling the other method?

---

*[server/zanata-war/src/test/java/org/zanata/rest/service/raw/ResourceServiceRestITCase.java, line 54 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-Kop2xxClT2fKbqMxzEm:-Kop2xxD6rOFkS6PWptQ:b-djhrjq) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/test/java/org/zanata/rest/service/raw/ResourceServiceRestITCase.java#L54)):*
> ```Java
>     public void testPutGetResourceWithExtension() {
>         log.debug("put resource:" + sr.toString());
>         getSourceDocResource().putResourceWithDocId(sr, sr.getName(), new StringSet(
> ```

Are there any tests for the backwards-compatibility (idNoSlash) methods? We just need to test that they convert docIds correctly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java, line 191 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooL_mbrJIE-FPVGjhK:-KopLY7Y-TzOzK6MCFBH:boubf62) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java#L191)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

I agree that we should return an entity, even for failures, which matches the requested type (XML or JSON). Given that the error strings weren't really machine readable before, I don't think backwards compatibility will be an issue.

This looks like a good pattern to follow, in terms of Enunciate annotations for error responses :-)

https://github.com/zanata/zanata-mt/blob/f8845de0fcd3ee05378ffe1d74dc31f2c247402c/server/src/main/java/org/zanata/mt/api/service/BackendResource.java#L37

</blockquote></details>

Aha! I found a couple of efforts to standardise error responses:

https://tools.ietf.org/html/rfc7807 "Problem Details for HTTP APIs" with JSON -`application/problem+json` - and XML -
 `application/problem+xml` - media types. I think we should use it. It should be straightforward to create a Problem DTO from the JSON and/or XML descriptions in the RFC.

There's also https://github.com/blongden/vnd.error which looks interesting, but there's been no activity for almost 3 years. I think an proposed standard RFC beats that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java, line 186 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooBykIHhqpOC5gUCSB:-KopZ89uAWVkRwh7kXFE:b-m8s6y1) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/api/zanata-common-api/src/main/java/org/zanata/rest/service/AsynchronousProcessResource.java#L186)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Sorry, I forgot this was a new end-point when I replied to Patrick. Yes, the DefaultValue should be false for this one.
</blockquote></details>

In fact, we should just remove this parameter. Implicitly running copytrans is a bad idea. As Patrick said, we have CopyTransResource now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java, line 191 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooL_mbrJIE-FPVGjhK:-KpD6ciyiqyy1uznVtcc:b-iunn9f) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/SourceDocResourceService.java#L191)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Aha! I found a couple of efforts to standardise error responses:

https://tools.ietf.org/html/rfc7807 "Problem Details for HTTP APIs" with JSON -`application/problem+json` - and XML -
 `application/problem+xml` - media types. I think we should use it. It should be straightforward to create a Problem DTO from the JSON and/or XML descriptions in the RFC.

There's also https://github.com/blongden/vnd.error which looks interesting, but there's been no activity for almost 3 years. I think an proposed standard RFC beats that.
</blockquote></details>

I think this is general issue in all endpoints that return string. I will refactor it as seperate PR.  Putting todo at the moment.

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/TranslatedDocResourceService.java, line 142 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KooM_ayuq9fUfbW7Zm3:-KpD7TFQjRBV4SLDGQv6:b8mwiqt) ([raw file](https://github.com/zanata/zanata-platform/blob/ba2abeba7915cf6e394cc8db465551360eeafa84/server/zanata-war/src/main/java/org/zanata/rest/service/TranslatedDocResourceService.java#L142)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Same here. Not json friendly.
</blockquote></details>

Seperate PR to return problem dto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by huangp** [comment]
Reviewed 5 of 20 files at r3.
Review status: 7 of 22 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by seanf** [comment]
Reviewed 15 of 20 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[client/zanata-client-commands/src/main/java/org/zanata/client/commands/push/PushCommand.java, line 597 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KpDdSX3_uxgNiAg20go:-KpDdSX4OSCUz2dJ3UCr:bg8fcht) ([raw file](https://github.com/zanata/zanata-platform/blob/8ced84de936f133f103482b0aae04a14dab79136/client/zanata-client-commands/src/main/java/org/zanata/client/commands/push/PushCommand.java#L597)):*
> ```Java
>                                                 getOpts().getProj(), getOpts()
>                                                         .getProjectVersion(),
>                                                 srcDoc, extensions, docId);
> ```

I suppose we should remove the copyTrans property from PushOptions too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[client/zanata-client-commands/src/main/java/org/zanata/client/commands/push/PushCommand.java, line 597 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KpDdSX3_uxgNiAg20go:-KpEDUxF-Zi5CiO71tSy:b-xanddm) ([raw file](https://github.com/zanata/zanata-platform/blob/8ced84de936f133f103482b0aae04a14dab79136/client/zanata-client-commands/src/main/java/org/zanata/client/commands/push/PushCommand.java#L597)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

I suppose we should remove the copyTrans property from PushOptions too.
</blockquote></details>

So removing --copy-trans option entirely?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by aeng** [comment]
Review status: 11 of 23 files reviewed at latest revision, 7 unresolved discussions.

---

*[client/zanata-client-commands/src/main/java/org/zanata/client/commands/push/PushCommand.java, line 597 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KpDdSX3_uxgNiAg20go:-KpEGRDBUlLj1z3B6qGC:b-anp3g9) ([raw file](https://github.com/zanata/zanata-platform/blob/8ced84de936f133f103482b0aae04a14dab79136/client/zanata-client-commands/src/main/java/org/zanata/client/commands/push/PushCommand.java#L597)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

So removing --copy-trans option entirely?
</blockquote></details>

Need to keep the --copy-trans option for the time being since we are running it after push process in zanata-cli.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by huangp** [comment]
:lgtm: 

---

Reviewed 1 of 20 files at r3, 12 of 12 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-:-KpEJh13HN-A92evTQ8B:bg7igkv)*

**Comment by seanf** [comment]
Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/zanata/zanata-platform/412#-KpHzl4SEEnkXcug00re:-KpHzl4Tlx-4UBtVe0nc:b8735xh) (no related file):*
@aeng, it looks like codecov moved the wrong way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by huangp** [comment]
Reviewed 1 of 7 files at r5, 11 of 11 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/412)*

**Comment by aeng** [comment]
Jenkins built: https://zanata-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/github-zanata-org/job/zanata-platform/view/Pull%20Requests/job/PR-412/34/
